pkg/git/git_test.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package git
+
+import "testing"
+
+func TestValidateGitSSHURLFormat(t *testing.T) {
+	tests := []struct {
+		url  string
+		want bool
+	}{
+		{
+			url:  "git@github.com:user/project.git",
+			want: true,
+		},
+		{
+			url:  "https://github.com/user/project.git",
+			want: false,
+		},
+		{
+			url:  "http://github.com/user/project.git",
+			want: false,
+		},
+		{
+			url:  "git@127.0.0.1:user/project.git",
+			want: true,
+		},
+		{
+			url:  "https://127.0.0.1/user/project.git",
+			want: false,
+		},
+		{
+			url:  "http://127.0.0.1/user/project.git",
+			want: false,
+		},
+		{
+			url:  "ssh://user@host.xz:port/path/to/repo.git/",
+			want: true,
+		},
+		{
+			url:  "ssh://user@host.xz/path/to/repo.git/",
+			want: true,
+		},
+		{
+			url:  "ssh://host.xz:port/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "ssh://host.xz/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "ssh://user@host.xz/~user/path/to/repo.git/",
+			want: true,
+		},
+		{
+			url:  "ssh://host.xz/~user/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "ssh://user@host.xz/~/path/to/repo.git",
+			want: true,
+		},
+		{
+			url:  "ssh://host.xz/~/path/to/repo.git",
+			want: false,
+		},
+		{
+			url:  "git://host.xz/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "git://host.xz/~user/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "http://host.xz/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "https://host.xz/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "https://127.0.0.1/user/project.git",
+			want: false,
+		},
+		{
+			url:  "/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "file://~/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "user@host.xz:/path/to/repo.git/",
+			want: true,
+		},
+		{
+			url:  "host.xz:/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "ssh://host.xz/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "user@host.xz:~user/path/to/repo.git/",
+			want: true,
+		},
+		{
+			url:  "host.xz:~user/path/to/repo.git/",
+			want: false,
+		},
+		{
+			url:  "user@host.xz:path/to/repo.git",
+			want: true,
+		},
+		{
+			url:  "host.xz:path/to/repo.git",
```

Comment:
I think this is a valid SSH url for git in the "scp-like syntax"?  `git` assumes the `user@` portion is the currently logged-in user.  So if current user in container is `git` then this will be interpreted as `git@host.xz:path/to/repo.git`

Reply:
The url with `user@` should be a valid SSH url like the example above, but without `user@`, it's not a valid one.

Reply:
@FogDong that's @sbwsg's point. From a git perspective, `git clone host.xz:path/to/repo.git` is valid (it will use the current logged in user, or the user specified in the `.ssh/config`)

Reply:
Indeed, sorry, my explanation was not clear.  From https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols :

> $ git clone [user@]server:project.git
> [...] **if you don’t specify the optional username, Git assumes the user you’re currently logged in as.**